pycqed/instrument_drivers/physical_instruments/ZurichInstruments/ZI_HDAWG8.py

```diff
@@ -1,1128 +1,460 @@
-        self._dev.daq.setInt('/' + self._dev.device +
-                             '/awgs/*/enable', 1)
+                for cw in range(2**6):
+                    # If either codeword is zero, fill that waveform with zeros
+                    # and mark it as read-only
+                    if (cw & mask_0) == 0:
+                        self.set(zibase.gen_waveform_name(2*awg_nr+0, cw), numpy.zeros(self._default_waveform_length))
+                        self._set_readonly_waveform(2*awg_nr+0, cw)
+
+                    if ((cw & mask_1) >> 3) == 0:
+                        self.set(zibase.gen_waveform_name(2*awg_nr+1, cw), numpy.zeros(self._default_waveform_length))
+                        self._set_readonly_waveform(2*awg_nr+1, cw)
+            else:
+                raise zibase.ziValueError('Invalid value {} selected for cfg_codeword_protocol!'.format(self.cfg_codeword_protocol()))
 
-        # Turn on all outputs
-        self._dev.daq.setInt('/' + self._dev.device + '/sigouts/*/on', 1)
         # Disable all function generators
-        self._dev.daq.setInt('/' + self._dev.device +
-                             '/sigouts/*/enables/*', 0)
-        # when doing flux pulses, set everything to amp mode
+        for param in [key for key in self.parameters.keys() if re.match(r'sines_\d+_enables_\d+', key)]:
+            self.set(param, 0)
+        
+        # Set amp or direct mode
         if self.cfg_codeword_protocol() == 'flux':
+            # when doing flux pulses, set everything to amp mode
             for ch in range(8):
                 self.set('sigouts_{}_direct'.format(ch), 0)
                 self.set('sigouts_{}_range'.format(ch), 5)
-
-        # Switch all outputs into direct mode when not using flux pulses
         else:
+            # Switch all outputs into direct mode when not using flux pulses
             for ch in range(8):
                 self.set('sigouts_{}_direct'.format(ch), 1)
                 self.set('sigouts_{}_range'.format(ch), .8)
 
-    def _find_valid_delays(self, awg, expected_sequence, verbose=False):
+        # Turn on all outputs
+        for param in [key for key in self.parameters.keys() if re.match(r'sigouts_\d+_on', key)]:
+            self.set(param, 1)
+
+    def _debug_report_dio(self):
+        # FIXME: only DIO 0 for now
+        log.info('DIO bits with timing errors:  0x%08X' % self.geti('awgs/0/dio/error/timing'))
+        log.info('DIO bits detected high:       0x%08X' % self.geti('awgs/0/dio/highbits'))
+        log.info('DIO bits detected low:        0x%08X' % self.geti('awgs/0/dio/lowbits'))
+        # AWGS/0/DIO/ERROR/WIDTH
+        # AWGS/0/DIO/DATA
+
+    # override for InstrumentBase
+    def snapshot_base(self, update=False, params_to_skip_update=None):
+        if params_to_skip_update is None:
+            params_to_skip_update = self._params_to_skip_update
+        snap = super().snapshot_base(
+            update=update, params_to_skip_update=params_to_skip_update)
```

Comment:
Although we overwrite the snapshot_base  here the more up to date version that was deleted is required.

Reply:
Done.